sigstore/_internal/oidc/ambient.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Ambient OIDC credential detection for sigstore.
+"""
+
+import logging
+import os
+from typing import Callable, List, Optional
+
+import requests
+from pydantic import BaseModel
+
+from sigstore._internal.oidc import IdentityError
+
+logger = logging.getLogger(__name__)
+
+
+class AmbientCredentialError(IdentityError):
+    """
+    Raised when an ambient credential should be present, but
+    can't be retrieved (e.g. network failure).
+    """
+
+    pass
+
+
+def detect_credential() -> Optional[str]:
```

Comment:
We don't need to address this in this PR, but I think `Optional[Identity]` would be a better signature here -- that way we could assert our validity guarantees about the token (e.g. having the right audience and being issued by a party we know).

Thoughts @tetsuo-cpp?

Reply:
(If we were to do this, we should also refactor our top-level `sign` API to take an `Identity` instead of constructing one internally.)

Reply:
Yeah I think that's a good idea. I'm going to take a stab at #4 in the coming days so it might be good to tackle it then so we get the interfaces right.